File: lab5/Dictionary.c
// Dictionary.c
// Shaan Saini
// shksaini
// lab5
// C Source file for testing the Dictionary ADTs


#include<stdio.h>
#include<stdlib.h>
#include<assert.h>

#include "Dictionary.h"

typedef struct Node{
  int key;
  int value;
  struct Node* next;
} Node;

typedef Node* NodeRef;

struct Dictionary{
  NodeRef front;
  int numItems;
}Dictionary;

NodeRef newNode(int x, int y){
  NodeRef N = malloc(sizeof(Node));
  assert(N!=NULL);
  N->key = x;
  N->value = y;
  N->next = NULL;
  return(N);
}

void freeNode(NodeRef* pN){
  if(pN!=NULL && *pN!=NULL){
    free(*pN);
    *pN = NULL;
  }
}

DictionaryRef newDictionary(void){
  DictionaryRef D = malloc(sizeof(Dictionary));
  assert(D!=NULL);
  D->front = NULL;
  D->numItems = 0;
  return D;
}

void freeDictionary(DictionaryRef* pD){
  free(*pD);
  *pD = NULL;
}

int isEmpty(DictionaryRef D){
  if( D == NULL){
    fprintf(stderr,"Stack Error: calling isEmpty() on NULL StackRef\n");
    exit(EXIT_FAILURE);
  }
  return(D->numItems==0);
}

int size(DictionaryRef D){
  return(D->numItems);
}

int lookup(DictionaryRef D, int k){
  NodeRef N=D->front;
  while(N != NULL){
    if(N->key == k){
      return N->value;
    }
    N = N->next;
  }
  return UNDEF;
}

void insert(DictionaryRef D, int x, int y){
  NodeRef N=D->front;
  if(D==NULL){
    fprintf(stderr, "Stack Error: calling insert() on NULL StackRef\n");
    exit(EXIT_FAILURE);
  }
  else if(lookup(D,y)!=UNDEF){
    fprintf(stderr, "Stack Error: inserting dublicate key\n");
    exit(EXIT_FAILURE);
  }
  else{
    if(D->front == NULL){
      N = newNode(x,y);
      D->front = N;
      D->numItems++;
    }
    else{
      while(N!= NULL){
	if(N->next==NULL){
	  break;
	}
	N = N->next;
      }
      N->next = newNode(x,y);
      D->numItems++;
    }
  }
}

void delete(DictionaryRef D, int k){
  NodeRef N = D->front;
  if(D==NULL){
    fprintf(stderr, "Stack Error: calling delete() on NULL StackRef\n");
    exit(EXIT_FAILURE);
  }
  else if(lookup(D,k)==UNDEF){
    fprintf(stderr, "Stack Error: deleting non-existant key\n");
    exit(EXIT_FAILURE);
  }
  else{
    if(D->numItems<=1){;
      freeNode(&N);
    }
    else{
      if(N->key==k){
	NodeRef temp = D->front; 
	D->front = N->next;
	D->numItems--;
	freeNode(&temp);
      }
      else{
	while(N->next->key!=k){
	  N = N->next;
	}
	NodeRef temp =  N->next;
	N->next = N->next->next;
	freeNode(&temp);
	D->numItems--;
      }
    }
  }
}

void makeEmpty(DictionaryRef D){
  D->numItems=0;
  freeNode(&D->front);
}

void printDictionary(DictionaryRef D, FILE* out){
  NodeRef N = D->front;
  while( N != NULL){
    fprintf(out,"%d %d \n",N->key,N->value);
    N = N->next;
  }
}

File: pa3/Dictionary.java
// Dictionary.java
// Shaan Saini
// shksaini
// pa3
// Dictionary class containing the method for the Dictionary ADT

public class Dictionary implements DictionaryInterface{
  
    private class Node{
	String key;
	String value;
	Node next;

	Node(String key, String value){
	    this.key = key;
	    this.value = value;
	    next = null;
	}
    }

    private Node front;
    private int numItems;

    public Dictionary(){
	front = null;
	numItems = 0;
    }
 
    public boolean isEmpty(){
	return (numItems==0);
    }

    public int size(){
	return numItems;
    }

    public String lookup(String key){
	Node N = front;
	while( N != null){
	    if( N.key.equals(key)){
		return N.value;
	    }
	    N = N.next;
	}
	return null;
    }

    public void insert(String key, String value) throws DuplicateKeyException{
	if( lookup(key) != null){
	    throw new DuplicateKeyException("cannot insert duplicate key");
	}
	else{
	    if( front == null){
		Node N = new Node(key,value);
		front = N;
		numItems++;
	    }
	    else{
		Node N = front;
		while( N != null){
		    if(N.next == null){
			break;
		    }
		    N = N.next;
		}
		N.next = new Node(key,value);
		numItems++;
	    }
	}
    }

    public void delete(String key) throws KeyNotFoundException{
	if( lookup(key) == null){
	    throw new KeyNotFoundException("cannot delete non-existent key");
	}
	else{
	    if(numItems <= 1){
		Node N = front;
		front = front.next;
		N.next = null;
		numItems--;
	    }
	    else{
		Node N = front;
		if(N.key.equals(key)){
		    front = N.next;
		    numItems--;
		}
		else{
		    while(!N.next.key.equals(key)){
			N = N.next;
		    }
		    N.next = N.next.next;
		    numItems--;
		}
	    }
	}
    }
    /*makeEmpty
     *Pre:none 
     *Pos:makes the linked list empty 
     */ 
    public void makeEmpty(){
	front = null;
	numItems =0;
    }

    public String toString(){
	String s = "";
	Node N = front;
	while( N != null){
	    s += N.key + " " + N.value + "\n"; 
	    N = N.next;
	}
	return s;
    }

    private Node findKey(String key){
	Node N = front;
	while(N != null){
	    if(N.key.equals(key)){
		return N; 
	    }
	    else{
		N = N.next; 
	    }
	}
	return null;
    }
  
}
